vowpalwabbit/parse_example.cc

```diff
@@ -50,29 +50,113 @@ int read_features_string(vw* all, v_array<example*>& examples)
   return static_cast<int>(num_chars_initial);
 }
 
+class feature_group_helper
+{
+public:
+  using char_map = std::array<unsigned char, NUM_NAMESPACES>;
+  using affix_map = std::array<uint64_t, NUM_NAMESPACES>;
+  using bool_map = std::array<bool, NUM_NAMESPACES>;
+  using namespace_dictionaries = std::vector<std::shared_ptr<feature_dict>>;
+  using dictionaries_map = std::array<namespace_dictionaries, NUM_NAMESPACES>;
+
+  feature_group_helper(example* current_example, bool redefine_char_flag, char_map* redefine_char_map,
+      affix_map* affix_features, bool_map* spelling_features, dictionaries_map* namespace_dictionaries_map,
+      uint32_t hash_seed, hash_func_t hash_fn)
+      : _feature_group_char(0)
+      , _hash(0)
+      , _hash_seed(hash_seed)
+      , _p_current_feature_group(nullptr)
+      , _p_affix_feature_group(nullptr)
+      , _p_spelling_feature_group(nullptr)
+      , _p_dictionary_feature_group(nullptr)
+      , _current_example(current_example)
+      , _redefine_char_flag(redefine_char_flag)
+      , _redefine_char_map(redefine_char_map)
+      , _namespace_dictionaries_map(namespace_dictionaries_map)
+      , _affix_features(affix_features)
+      , _spelling_features(spelling_features)
+      , _hash_function(hash_fn)
+  {
+  }
+  unsigned char feature_group_char() const { return _feature_group_char; }
+  uint64_t feature_group_hash() const { return _hash; }
+  features& get_current_fg() { return get_feature_group(_p_current_feature_group, feature_group_char()); }
+  features& get_affix_fg() { return get_feature_group(_p_affix_feature_group, affix_namespace); }
+  features& get_spelling_fg() { return get_feature_group(_p_spelling_feature_group, spelling_namespace); }
+  features& get_dictionary_fg() { return get_feature_group(_p_dictionary_feature_group, dictionary_namespace); }
+  void set_current_fg(const VW::string_view fg_name)
```

Comment:
nit:

since `fg_name` is passed by value I think `const` is redundant

Reply:
const tells the compiler to enforce that fg_name won't change inside the function and perhaps allows opportunities for optimization?

Reply:
but since it's passed by value it does it matter if it changes inside the function, the guarantee stays the same from the caller point of view afaik

Reply:
I guess it doesn't matter either way (hence the nit)

Reply:
When the argument is a value anyway there is little reason to make it const. The compiler isn't really going to do anything extra because of the const (and it rarely does because const in C++ isnt very binding...).